### ISSUE TITLE
How does this model handle priors cross image boundary?

### ISSUE THREAD

**Issue by placeforyiming** [kujason/avod#138]
In faster r-cnn, all cross-boundary priors are discarded during the training. Dose anyone know how AVOD process those 3D priors which cross image boundary after mapping?

I read the paper carefully, but can't find any place address this problem. Any suggestions are very helpful for me. Thanks!

**Comment by placeforyiming** [comment]
I believe this is an important point, since the faster r-cnn paper claims those priors will make the model fail to converge.  But if we discard 3D priors, we can't handle those ground truths partially appear on image.

**Comment by jhultman** [comment]
The Faster-RCNN [paper](https://arxiv.org/abs/1506.01497) mentioned:
> If the boundary-crossing outliers are not ignored in training, they introduce large, difficult to correct error terms in the objective, and training does not converge.

But Ross Girshick has since [suggested](https://github.com/facebookresearch/Detectron/issues/393#issuecomment-384660309) that this is perhaps less of an issue than was once believed:
> It seems that this choice does not really matter. I have tried both (for a few limited models) and found very similar performance. The original observation from the Faster R-CNN and FPNs paper may have been due to implementation issues, not fundamental problems.

So the quote from the original paper probably should not be taken too literally.

**Comment by placeforyiming** [comment]
@jhultman Thanks!

However, in 3D case, we need fuse the cropped region from the image feature map with it's counterpart from lidar feature map. If an object fully appear on image, both the lidar and image contain the same information. But if an object partially appear on image, the fusion is inconsistent. 

I'm trying to pad zero for the partial car in my own version of AVOD. It takes longer for the model to detect cars on boundary. But the final performance is ok, the model only miss cars which show a small part on image.

**Comment by jhultman** [comment]
@placeforyiming Ah yes, you are right. I didn't consider that the projected boxes are used to crop and extract features from the image. But I think the authors have thought about this even though it may not be explicitly mentioned in their paper. [This](https://github.com/kujason/avod/blob/b2d32f6ddd5007c12afe37760fbcff993816da39/avod/core/box_3d_projector.py#L142) is the code that handles anchors that cross image boundaries. The default argument is `truncate=False` and when called in `rpn_model.py` this argument is not overwritten, which means the projected boxes may extend past image boundaries. 

If we look at [`tf.image.crop_and_resize`](https://www.tensorflow.org/api_docs/python/tf/image/crop_and_resize) which does the ROI pooling for the image features, the behavior is to use `extrapolation_value=0` for crop regions extending beyond the image. So I think the default behavior is to zero pad before bilinear interpolation in ROI pooling op. I think you can make a good argument for truncating the projected boxes, since zero-padding will bias the pooled features toward zero for cross-boundary proposals, but maybe the authors have tried both ways and determined it had negligible effect.

**Comment by placeforyiming** [comment]
Thanks for discussion!
